### PR TITLE
✨Amend preloading to work with branching

### PR DESCRIPTION
**Pull request by bramanudom**
#20083

This PR adds additional logic make preloading consider next pages that are reachable within stories through `goToPage` action, as well as previous pages that were reached through either `advance-to` or `goToPage`.

For next pages we search through the page element to find the next pages. For previous pages, we make use of the story level stack (#20182).

### CODE REVIEW

File: extensions/amp-story/1.0/amp-story.js [hunk 1]
```diff
@@ -1799,11 +1799,29 @@ export class AmpStory extends AMP.BaseElement {
...
+        const maybePrev = this.storyNavigationPath_[indexInStack - 1];
+        if (indexInStack > 0 && pageId === this.activePage_.element.id) {
+          if (!pagesByDistance[1]) {
+            pagesByDistance[1] = [];
+          }
+          pagesByDistance[1].push(maybePrev);
+        }
+        // Do not overwrite, branching distance always takes precedence.
+        if (pageId !== maybePrev) {
+          pagesByDistance[distance].push(pageId);
```
Comment (newmuis) [inline extensions/amp-story/1.0/amp-story.js]: Can this be combined with the statement below?

File: extensions/amp-story/1.0/amp-story-page.js [hunk 2]
```diff
@@ -767,6 +770,25 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+  getNextPagesByAction() {
+    const actionElements =
+      Array.from(
+          this.element.querySelectorAll('[on]'),
+          el => el.getAttribute('on'));
+
+    return actionElements.reduce((res, el) => {
+      if (el.includes('goToPage')) {
+        // The pageId is in between the equals sign & closing parenthesis.
+        res.push(el.slice(el.search('\=(.*)') + 1,-1));
```
Comment (cvializ) [inline extensions/amp-story/1.0/amp-story-page.js]: Will this still work if the element has two actions with parameters? e.g. 
```
on="tap:foo.barAction(baz=1), story.goToPage(id=page-id)"
```

File: extensions/amp-story/1.0/amp-story.js (hunk 1 above)
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story.js]: I think we'd still like to go through the step of setting the distance attribute when the branching experiment is turned off!

File: extensions/amp-story/1.0/amp-story-page.js (hunk 2 above)
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story-page.js]: Thank you for pointing this out! You're right, this won't work for that case.  It also won't work for handling multiple events either : `on="tap:story.goToPage(id=pageID);submit-error:lightbox2"`.

Made some changes to account for this.

File: extensions/amp-story/1.0/amp-story-page.js [hunk 3]
```diff
@@ -767,6 +770,25 @@ export class AmpStoryPage extends AMP.BaseElement {
     return null;
   }
 
+  /**
+   * Finds any elements in the page that has a goToPage action.
+   * @return {Array<string>} The IDs of the potential next pages in the story
```
[the remark this answers is not in the document]
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story-page.js]: Done

File: extensions/amp-story/1.0/amp-story-page.js [hunk 4]
```diff
@@ -767,6 +770,25 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+   * or [] if there aren't any.
+   */
+  getNextPagesByAction() {
+    const actionElements =
+      Array.from(
+          this.element.querySelectorAll('[on]'),
+          el => el.getAttribute('on'));
+
+    return actionElements.reduce((res, el) => {
+      if (el.includes('goToPage')) {
```
[the remark this answers is not in the document]
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story-page.js]: Done

File: extensions/amp-story/1.0/amp-story-page.js [hunk 5]
```diff
@@ -767,6 +770,25 @@ export class AmpStoryPage extends AMP.BaseElement {
     return null;
   }
 
+  /**
+   * Finds any elements in the page that has a goToPage action.
+   * @return {Array<string>} The IDs of the potential next pages in the story
+   * or [] if there aren't any.
+   */
+  getNextPagesByAction() {
+    const actionElements =
```
[the remark this answers is not in the document]
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story-page.js]: Done

File: extensions/amp-story/1.0/amp-story-page.js (hunk 2 above)
Comment (newmuis) [reply extensions/amp-story/1.0/amp-story-page.js]: Maybe we should also add unit tests for those cases?

Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story-page.js]: Good point. I just added some tests to test out the method I wrote.

File: extensions/amp-story/1.0/amp-story-page.js [hunk 6]
```diff
@@ -767,6 +770,29 @@ export class AmpStoryPage extends AMP.BaseElement {
     return null;
   }
 
+  /**
+   * Finds any elements in the page that has a goToPage action.
+   * @return {!Array<string>} The IDs of the potential next pages in the story
+   * or null if there aren't any.
+   */
+  actions() {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js]: Nit: should be private

File: extensions/amp-story/1.0/test/test-amp-story.js [hunk 7]
```diff
@@ -1388,6 +1388,47 @@ describes.realWin('amp-story', {
...
+
+            const actionButton = createElementWithAttributes(
+                win.document,
+                'button',
+                {'id': 'actionButton',
+                  'on': 'tap:story.goToPage(id=page-2)'});
+
+            story.element.querySelector('#cover').appendChild(actionButton);
+
+            const distanceGraph = story.getPagesByDistance_();
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/test/test-amp-story.js]: Nit: Can we set the goToPage action in the page before the buildCallback, so we don't have to manually call this private method?

File: extensions/amp-story/1.0/amp-story-page.js [hunk 8]
```diff
@@ -767,6 +770,29 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+    const actionElements =
+      Array.prototype.slice.call(this.element.querySelectorAll('[on]'));
```
[the remark this answers is not in the document]
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story-page.js]: Thank you for linking the documentation also, very good to know. Went ahead and did the change. I believe the inside check in the reduce for 'goToPage' still needs to be there to account for the cases outlined above (multiple events/actions).

File: extensions/amp-story/1.0/amp-story-page.js (hunk 6 above)
Comment (bramanudom) [reply extensions/amp-story/1.0/amp-story-page.js]: Done.

File: extensions/amp-story/1.0/test/test-amp-story.js (hunk 7 above)
Comment (bramanudom) [reply extensions/amp-story/1.0/test/test-amp-story.js]: I had some issues retrieving the distance attribute from the pages for some reason (it doesn't seem to exist during the tests on the page elements during the tests, though it's definitely being set). If it's okay with you, I think we should keep the as they are.

**Comment by newmuis** [comment]
@bramanudom is this ready to be merged?

**Comment by bramanudom** [comment]
@newmuis Yes! Thanks, Jon.

**Pull request merged**